### PR TITLE
Rename `input_type` --> `input_types` on AggregateFunctionExpr / AccumulatorArgs / StateFieldsArgs

### PR DESCRIPTION
**Pull request by lewiszlw**
## Which issue does this PR close?

Closes #.

## Rationale for this change

It confused me when I read these code that AggregateFunctionExpr / AccumulatorArgs / StateFieldsArgs only contain one input type.

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Thanks @lewiszlw  -- even though this is an API change I agree it makes the code less confusing

**Comment by jcsherin** [comment]
> It confused me when I read these code that AggregateFunctionExpr / AccumulatorArgs / StateFieldsArgs only contain one input type.

This confused me too the first time when I looked at code in user-defined aggregates. I thought the reason it was written like that was because for all existing user-defined aggregates knowing the type of the first argument is enough.

**Comment by jcsherin** [comment]
@lewiszlw It will be very helpful if you could please also update this example. https://github.com/apache/datafusion/blob/204e1bcbaa5fcd3cf3cbe045f9832ee2f669e92f/datafusion/functions-aggregate/COMMENTS.md?plain=1#L57 

I checked your branch and this file doesn't exist. So you will need to merge from main to get it.

**Comment by lewiszlw** [comment]
Thanks for pointing out. I'll update pr in a few days.

**Comment by alamb** [comment]
I happened to have this PR opened locally (I get anxious with PRs that are open too long 😅 ) so I took the liberty of updating the docs as well in 1a3c5ca7b while I was merging up from main

File: datafusion/functions-aggregate/COMMENTS.md [hunk 1]
```diff
@@ -54,7 +54,7 @@ first argument and the definition looks like this:
 // `input_type` :  data type of the first argument
 let mut fields = vec![Field::new_list(
     format_state_name(self.name(), "nth_value"),
-    Field::new("item", args.input_type.clone(), true /* nullable of list item */ ),
+    Field::new("item", args.input_types.clone(), true /* nullable of list item */ ),
```
Comment (jcsherin) [inline datafusion/functions-aggregate/COMMENTS.md:57]: Thanks @alamb for doing this 🙏.

The second argument should be `args.input_types[0].clone` according to the changes in this PR.

Comment (alamb) [reply datafusion/functions-aggregate/COMMENTS.md:57]:
```suggestion
    Field::new("item", args.input_types[0].clone(), true /* nullable of list item */ ),
```

**Comment by jayzhan211** [comment]
I think we don't even need vector of input type, since we only get the first one. 🤔

**Comment by alamb** [comment]
> I think we don't even need vector of input type, since we only get the first one. 🤔

Wouldn't we need a vector of input types if the aggregate had more than one argument? For example

https://docs.rs/datafusion/latest/datafusion/functions_aggregate/approx_percentile_cont/fn.approx_percentile_cont.html

**Comment by jayzhan211** [comment]
> > I think we don't even need vector of input type, since we only get the first one. 🤔
> 
> Wouldn't we need a vector of input types if the aggregate had more than one argument? For example
> 
> https://docs.rs/datafusion/latest/datafusion/functions_aggregate/approx_percentile_cont/fn.approx_percentile_cont.html

Second argument is fixed type (f64), so we don't have the actual function that expect different multiple input type yet.
Btw, I think we could remove `input_type`, because we could compute the type from `expr` and `schema`.
If there are any UDAF expect more flexible accumulator based on input types, they can get the types from `expr` and `schema` too.

Therefore, I recommend we remove input_type entirely and compute only if we need it in `create_accumulator`

**Comment by alamb** [comment]
> Therefore, I suggest we remove input_type entirely and compute only if we need it in create_accumulator from input_exprs and schema. This not only make the code less confused and also let udaf be flexible and efficient (only compute types if needed)

That certainly sounds cleaner.

I feel like we have churned this API a bunch recently, so maybe we can take a step back and ensure that whatever we come up with supports the usecases we know of now (and in the future) so we don't keep changing it 🤔

**Comment by jayzhan211** [comment]
> > Therefore, I suggest we remove input_type entirely and compute only if we need it in create_accumulator from input_exprs and schema. This not only make the code less confused and also let udaf be flexible and efficient (only compute types if needed)
> 
> That certainly sounds cleaner.
> 
> I feel like we have churned this API a bunch recently, so maybe we can take a step back and ensure that whatever we come up with supports the usecases we know of now (and in the future) so we don't keep changing it 🤔

I plan to change `input_exprs` to `physical-exprs` in #11359, so there will be another change.

I think it would be nice to have something like `unstable feature` (maybe declare as part of the docs) similar to `nightly` in rust for feature or APIs that takes over one month (release cycle). After most of the things are done, we could consider it `stable` feature. We still do breaking change for `stable` feature, but it gives clear message to others about `unstable` feature that is in frequent changing stage, so there could be less impact on downstream, and also easier for developer to achieve the better design

**Comment by jayzhan211** [comment]
The end state in my mind now.
```rust
pub struct AccumulatorArgs<'a> {
    /// Keep
    pub data_type: &'a DataType,

    /// We might only need one of `schema` or `dfschema`. It is likely we keep `dfschema`, since we can get `schema` from it.
    pub schema: &'a Schema,

    pub dfschema: &'a DFSchema,

    /// Keep
    pub ignore_nulls: bool,

    /// Convert to physical sort exprs instead
    pub sort_exprs: &'a [Expr],

    /// Keep
    pub is_reversed: bool,

    /// We might able to get the name from expressions
    pub name: &'a str,

    /// Keep
    pub is_distinct: bool,

    /// Get the type for schema and expressions
    pub input_type: &'a DataType,

    /// Convert to physical expressions
    pub input_exprs: &'a [Expr],
}
```

**Comment by alamb** [comment]
> ```rust
> pub struct AccumulatorArgs<'a> {
>     /// Keep, this is return type, the name might be quite confusing.
>     pub data_type: &'a DataType,

I agree `return_type` would be clearer

> 
>     /// We might only need one of `schema` or `dfschema`. It is likely we keep `dfschema`, since we can get `schema` from it.

I think `schema` is what makes sense as `dfschema` is mostly a logical thing in my mind (as it has a relation name). 

Further evidence that we don't need dfschema is  that you can get a `DataType` from a `PhysicalExpr` using a `Schema` https://docs.rs/datafusion/latest/datafusion/physical_expr/trait.PhysicalExpr.html#tymethod.data_type

> 
>     /// Convert to physical sort exprs instead
>     pub sort_exprs: &'a [Expr],

Agree

> 
>     /// Convert to physical expressions
>     pub input_exprs: &'a [Expr],
> }

Make sense


Thanks @jayzhan211

**Comment by alamb** [comment]
So what should we do with this PR? Merge it and then revamp in a follow on PR?

**Comment by jayzhan211** [comment]
Sure, but it is better to remove `input_types` before next release @lewiszlw

**Comment by alamb** [comment]
> Sure, but it is better to remove `input_types` before next release @lewiszlw

Filed https://github.com/apache/datafusion/issues/11725 to track the cleanup

Thanks agin

**Comment by jayzhan211** [comment]
> I think `schema` is what makes sense as `dfschema` is mostly a logical thing in my mind (as it has a relation name).

@xinlifoobar I think it is not clear whether we should keep `dfschema` or `schema`, so I suggest we just remove `input_types` at this moment. After the refactor #11359 , it would be clear

**Pull request merged**